ksqldb-functional-tests/src/test/resources/query-validation-tests/formats.json

```diff
@@ -17,6 +17,265 @@
+        "message": "The use of 'KEY_FORMAT' and 'FORMAT' is disabled, as this feature is under development."
+      },
+      "inputs": [],
+      "outputs": []
+    },
+    {
+      "name": "format without feature flag",
+      "statements": [
+        "CREATE STREAM TEST (foo VARCHAR) WITH (kafka_topic='test_topic', format='KAFKA');"
+      ],
+      "topics": [
+        {
+          "name": "test_topic",
+          "format": "KAFKA"
+        }
+      ],
+      "expectedException": {
+        "type": "io.confluent.ksql.util.KsqlStatementException",
+        "message": "The use of 'KEY_FORMAT' and 'FORMAT' is disabled, as this feature is under development."
+      },
+      "inputs": [],
+      "outputs": []
+    },
+    {
+      "name": "explicit key and optional value format",
+      "statements": [
+        "CREATE STREAM INPUT (K STRING KEY, foo INT) WITH (kafka_topic='input_topic', key_format='KAFKA');",
+        "CREATE STREAM OUTPUT AS SELECT * FROM INPUT;"
+      ],
+      "properties": {
+        "ksql.persistence.default.format.value": "JSON",
+        "ksql.key.format.enabled": true
+      },
+      "inputs": [
+        {"topic": "input_topic", "value": {"FOO": 10}}
+      ],
+      "outputs": [
+        {"topic": "OUTPUT", "value": {"FOO": 10}}
+      ]
+    },
+    {
+      "name": "optional key and value format",
+      "statements": [
+        "CREATE STREAM INPUT (K STRING KEY, foo INT) WITH (kafka_topic='input_topic');",
+        "CREATE STREAM OUTPUT AS SELECT * FROM INPUT;"
+      ],
+      "properties": {
+        "ksql.persistence.default.format.value": "JSON",
+        "ksql.key.format.enabled": true
+      },
+      "inputs": [
+        {"topic": "input_topic", "value": {"FOO": 10}}
+      ],
+      "outputs": [
+        {"topic": "OUTPUT", "value": {"FOO": 10}}
+      ]
+    },
+    {
+      "name": "unsupported key format",
+      "statements": [
+        "CREATE STREAM TEST (foo VARCHAR) WITH (kafka_topic='test_topic', value_format='JSON', key_format='AVRO');"
+      ],
+      "properties": {
+        "ksql.key.format.enabled": true
+      },
+      "topics": [
+        {
+          "name": "test_topic",
+          "format": "JSON"
+        }
+      ],
+      "expectedException": {
+        "type": "io.confluent.ksql.util.KsqlStatementException",
+        "message": "The key format 'AVRO' is not currently supported."
+      },
+      "inputs": [],
+      "outputs": []
+    },
+    {
+      "name": "invalid key format",
+      "statements": [
+        "CREATE STREAM TEST (foo VARCHAR) WITH (kafka_topic='test_topic', value_format='JSON', key_format='FOO');"
+      ],
+      "properties": {
+        "ksql.key.format.enabled": true
+      },
+      "topics": [
+        {
+          "name": "test_topic",
+          "format": "JSON"
+        }
+      ],
+      "expectedException": {
+        "type": "io.confluent.ksql.util.KsqlStatementException",
+        "message": "Unknown format: FOO"
+      },
```

Comment:
This error is thrown by `FormatFactory`. We could pipe in information about whether the format being created is a key or value format in order to improve this error message but it's not clear to me how worthwhile this is. Thoughts?

Reply:
I guess if we validate the config on start up, so that the user sees an explicit error message such as:

```
Invalid format: FOO set in config ksql.persistence.key.format.default
```

... or similar, then statements will only fail if the user _explicitly_ supplies an invalid format in the statement, meaning the source of the error should be obvious!